### PR TITLE
Multi Value Slice

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Design doc: https://www.notion.so/arbitrum/A-multi-value-slice-f487fabde5f541a2a241a0b05bd9197b

### CODE REVIEW

File: container/multi-value-slice/multi_value_slice.go [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/prysmaticlabs/prysm/v4/container/multi-value-slice/interfaces"
+)
+
+// MultiValueSlice defines an abstraction over all concrete implementations of the generic Slice.
+type MultiValueSlice[O interfaces.Identifiable] interface {
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:11]: Perhaps `SharedValueSlice` is a better name for this

File: container/multi-value-slice/multi_value_slice.go [hunk 2]
```diff
@@ -0,0 +1,380 @@
+package multi_value_slice
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:1]: It would be nice to add a package level godoc comment explaining this feature and why we need it. It should also explain a bit about how to use it. How should I initialize a multi value slice ? What should I be aware of when using it? When should I clean it up or destroy it?

File: container/multi-value-slice/multi_value_slice.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	Len(obj O) int
+}
+
+// Value defines a single value along with one or more IDs that share this value.
+type Value[V any] struct {
+	val V
+	ids []interfaces.Id
+}
+
+// MultiValue defines a collection of Value items.
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:21]: Im not as a clear in the difference between this MultiValue type and the Slice type when it comes to naming. Maybe we need a better way to differentiate them. Perhaps, `SharedValueManager` instead of `Slice` ?

File: container/multi-value-slice/multi_value_slice.go [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+}
+
+// Len returns the number of items for the input object.
+func (s *Slice[V, O]) Len(obj O) int {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	l, ok := s.cachedLengths[obj.Id()]
+	if !ok {
+		return len(s.sharedItems)
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply container/multi-value-slice/multi_value_slice.go:55]: +1

File: container/multi-value-slice/multi_value_slice.go [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+// Copy copies items between the source and destination.
+func (s *Slice[V, O]) Copy(src O, dst O) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, item := range s.individualItems {
+		for _, v := range item.Values {
+			found := compareIds(v.ids, src.Id(), func(_ int) {
+				v.ids = append(v.ids, dst.Id())
+			})
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:69]: Not sure I fully understand this function call. Seems like found is a boolean, but compareIds has a side effect of appending? Would be good to either rename or clarify with commentary

File: container/multi-value-slice/multi_value_slice.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+					break
+				}
+			}
+			if !found {
+				result[i] = item
+			}
+		}
+	}
+
+	for _, item := range s.appendedItems {
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:126]: There's a lot of looping over items, values, doing compare and break if found. This could be abstracted into its own helper function that could be reused several times in this package

File: container/multi-value-slice/multi_value_slice.go [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+			return result
+		}
+	}
+
+	return result
+}
+
+// At returns the item at the requested index for the input object.
+// If the object has an individual value at that index, it will be returned. Otherwise the shared value will be returned.
+// If the object has an appended value at that index, it will be returned.
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:148]: Something im unclear about is the ordering of return types: why do we return individual first, then shared, then appended in that order?

File: container/multi-value-slice/multi_value_slice.go [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+	if index >= uint64(len(s.sharedItems)+len(s.appendedItems)) {
+		var def V
+		return def, fmt.Errorf("index %d out of bounds", index)
+	}
+
+	isOriginal := index < uint64(len(s.sharedItems))
+	if isOriginal {
+		ind, ok := s.individualItems[index]
+		if !ok {
+			return s.sharedItems[index], nil
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:162]: What if it's also not in the shared items mapping?

File: container/multi-value-slice/multi_value_slice.go [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+		}
+		for _, v := range ind.Values {
+			for _, id := range v.ids {
+				if id == obj.Id() {
+					return v.val, nil
+				}
+			}
+		}
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:170]: Duplicate for loop logic below

File: container/multi-value-slice/multi_value_slice.go [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+			// then all larger indices don't have an appended item for the object either.
+			break
+		}
+	}
+
+	delete(s.cachedLengths, obj.Id())
+}
+
+// compareIds executes a custom action when the wanted ID is found. It returns whether the ID was found.
+func compareIds(ids []interfaces.Id, wanted interfaces.Id, action func(index int)) bool {
```
Comment (rauljordan) [inline container/multi-value-slice/multi_value_slice.go:370]: The side-effect callback is a bit hard to follow when using this. Why not just compare ids and return whether it was found? and then do the action afterwards from the caller

Comment (rauljordan) [reply container/multi-value-slice/multi_value_slice.go:370]: Maybe should be called `containsId`

File: container/multi-value-slice/multi_value_slice.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go:55]: We only cache lengths for objects that have at least one appended item to avoid looping through appended items every time we call `Len`.

File: container/multi-value-slice/multi_value_slice.go [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+		for vi, v := range ind.Values {
+			found := compareIds(v.ids, obj.Id(), func(index int) {
+				if len(v.ids) == 1 {
+					if len(ind.Values) == 1 {
+						delete(s.individualItems, i)
+					} else {
+						ind.Values = append(ind.Values[:vi], ind.Values[vi+1:]...)
+					}
+				} else {
+					v.ids = append(v.ids[:index], v.ids[index+1:]...)
```
[the remark this answers is not in the document]
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go:336]: Great catch! This is an obvious bug.

File: container/multi-value-slice/multi_value_slice.go (hunk 5 above)
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go:69]: We look for the Id of `src` and append the Id of `dst`

File: container/multi-value-slice/multi_value_slice.go (hunk 3 above)
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go:21]: I renamed `MultiValue` to `MultiValueItem`. I wouldn't call this a `Manager` because after all it's just a container.

File: container/multi-value-slice/multi_value_slice.go (hunk 7 above)
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go:148]: Appended items' indices are always larger than shared/individual items' indices. We first check if the index is within the length of shared items. If it is, then we return an individual value at that index - if it exists - or a shared value otherwise. If the index is beyond the length of shared values, it is an appended item and that's what gets returned.

I will change the description to the above, which is more detailed. And more clear, I hope.

File: container/multi-value-slice/multi_value_slice.go (hunk 8 above)
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go:162]: We already check this here:
```
isOriginal := index < uint64(len(s.sharedItems))
```

If the index is within the length of shared items, it means we either have a shared value or an individual value. There's no third option.

File: container/multi-value-slice/multi_value_slice.go (hunk 10 above)
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go:370]: I removed the callback. I now return the found index from the function and use that index for subsequent actions.

File: container/multi-value-slice/multi_value_slice.go (hunk 2 above)
Comment (rkapka) [reply container/multi-value-slice/multi_value_slice.go]: Very good point. I will add something.

File: container/multi-value-slice/multi_value_slice.go [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+// and we append an item for object0 and then append another item for object1.
+// Now we have two items in the slice, but object1 only has an item in index 1. This makes things very confusing and hard to deal with.
+// If we make appended items a []*Value, things don’t become much better.
+// It is therefore easiest to make appended items a []*MultiValueItem, which allows each object to have its own values starting at index 0
+// and not having any “gaps”.
+//
+// The Detach function should be called when an object gets garbage collected.
+// Its purpose is to clean up the slice from individual/appended values of the collected object.
+// Otherwise the slice will get polluted with values for non-existing objects.
+package mvslice
```
Comment (rkapka) [inline container/multi-value-slice/multi_value_slice.go:37]: I made the package name more idiomatic. Removing underscores resulted in `multivalueslice`, which I deemed to long, so I came up with `mvslice`.

**Review by rauljordan — APPROVED** [review]

**Comment by rkapka** [comment]
Let's wait with merging until we finish testing the feature in the dev cluster.

File: container/multi-value-slice/multi_value_slice.go [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+		for _, v := range item.Values {
+			_, found := containsId(v.ids, src.Id())
+			if found {
+				v.ids = append(v.ids, dst.Id())
+				break
+			}
+		}
+	}
+
+appendedLoop:
```
Comment (nisdas) [inline container/multi-value-slice/multi_value_slice.go:112]: you don't need the goto label here, it doesn't serve any purpose.

File: container/multi-value-slice/multi_value_slice.go [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+	}
+}
+
+// Value returns all items for the input object.
+func (s *Slice[V, O]) Value(obj O) []V {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	result := make([]V, len(s.sharedItems))
+	for i, item := range s.sharedItems {
```
Comment (nisdas) [inline container/multi-value-slice/multi_value_slice.go:141]: this whole loop can be a separate method here

File: container/multi-value-slice/multi_value_slice.go [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+func (s *Slice[V, O]) UpdateAt(obj O, index uint64, val V) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if index >= uint64(len(s.sharedItems)+len(s.appendedItems)) {
+		return fmt.Errorf("index %d out of bounds", index)
+	}
+
+	isOriginal := index < uint64(len(s.sharedItems))
+	if isOriginal {
```
Comment (nisdas) [inline container/multi-value-slice/multi_value_slice.go:231]: isOriginal and its negative could be their own separate methods. The `UpdateAt` method is hard to read and on an initial look

**Review by nisdas — APPROVED** [review]
As a follow up we should start fuzzing this and comparing the results with this vs a normal slice to ensure correctness
